ADR/0039-workspace-deprecation.md

```diff
@@ -0,0 +1,104 @@
+
+* We want to make Konflux a successful open source projects that fits into the
+cloud native apps ecosystem. We want everything to work on vanilla Kubernetes - no OpenShift required.
+
+* We had our [first upstream Konflux release](https://github.com/konflux-ci/konflux-ci) which doesn't use [kubesaw] or relying
+on the Workspace abstraction, and is deployable on vanilla Kubernetes.
+
+* We made a decision (now long ago) to drop [KCP] from our architecture, which initially introduced the workspace concept. Both transparent multi-cluster as a feature and workspaces as a concept came from KCP.
+
+* We made the decision (at Red Hat) to stop trying to offer a managed service interface for Konflux. The user signup, waitlist, and workspace provisioning interfaces made a lot of sense there.
+
+## Decision
+
+1. We will remove the Workspace abstraction from Konflux and use standard Kubernetes namespaces
+directly for providing a place for users to run their pipelines and store their configurations
+and secrets.
+
+2. Konflux won't provide its own policy engine for enforcing policies related to
+creating/deleting namespaces. It will instead recommend user to use existing open source projects
+such as [Kyverno](https://kyverno.io/docs/introduction/) and [Gatekeeper](https://open-policy-agent.github.io/gatekeeper/website/docs/) for this task.
+
+3. We will stop using the term Workspace, and start to use the term Namespace.
+
+4. The Konflux UI will expose a wizard for creating a new namespace. This
+wizard will be available to any user that has permissions to create namespaces.
+
+5. Konflux will provide an optional automation for creating a namespace for a user
+when he/she logins for the first time.
+
+6. Konflux won't be opinionated about the mechanism for initializing/maintaining namespaces with
+supporting resources such as ResourceQuota and LimitRange. Konflux will defer
+this responsibility to other tools which are specialized in this task such as 
+(but not limited to) [Kyverno](https://kyverno.io/policies/best-practices/add-ns-quota/add-ns-quota/) and [ArgoCD](https://github.com/konflux-ci/namespace-generator).
+
+7. Konflux will provide a thin [backend service](https://github.com/konflux-ci/workspace-manager) for listing the namespaces where the user has at least view access
+to the Konflux CRDs. This list will be used by the namespace switcher in the UI.
+This is required since the Kubernetes API doesn't let the user to list a subset
+of namespace. The user gets permissions to list all namespace or none.
+
+8. Konflux will provide ClusterRoles that will grant permissions to the Konflux
+and Tekton CRDs. Those will be (aggregated)[https://github.com/konflux-ci/konflux-ci/issues/440] to the built-in Kubernetes roles (`view`, `edit`, `admin`).
+
+9. Public viewer access will be provided by assigning the `view` role (see above) to the
+`system:authenticated` built-in group that contains all the authenticated users.
```

Comment:
Under this model, will private namespaces going to be supported?  If so:

1. We're going to need to have a mechanism for distinguishing between public and private namespaces, so that access can be given only to the appropriate users.  How does this model plan on implementing this mechanism?
2. We're also going to need to only give permissions to only view Konflux and Tekton resources in namespaces that a user has access to.  In RBAC terms, we need to provide the roles in each namespace, since a single ClusterRole across the entire cluster won't be granular enough.  How does this model propose to handle this case?

Reply:
Are you intending on pre-populating these role bindings into newly-made namespaces?  If so, this sounds like namespace templates, which this proposal is explicitly against doing.  If not, then what component will be responsible for ensuring namespaces have had these resources installed?

Reply:
| If not, then what component will be responsible for ensuring namespaces have had these resources installed?

I suggest that Konflux shouldn't be opinionated about it. I suggested some tools that can be used for that, but it's only a suggestion.

If the are no objections I'll close this thread.